tools/chpldef/Message.cpp

```diff
@@ -315,43 +329,64 @@ const char* Message::errorToString(Error error) {
   return nullptr;
 }
 
-namespace {
-template <typename T, typename X>
-static bool doHandleMessage(Server* ctx, T* msg, X& x) {
+template <typename M, typename P, typename R>
+static bool
+doHandleNotification(Server* ctx, M* msg, const P& p, R& r) {
+  if (msg->status() == Message::PROGRESSING) CHPLDEF_TODO();
+  if (msg->status() != Message::PENDING) return false;
+
+  ctx->message("Handling notification '%s'\n", msg->tagToString());
+
+  r = M::compute(ctx, p);
+
+  if (r.isProgressingCallAgain) CHPLDEF_TODO();
+
+  ctx->message("Notification complete...\n");
+
+  return true;
+}
+
+template <typename M, typename P, typename R>
+static bool
+doHandleRequest(Server* ctx, M* msg, const P& p, R& r) {
   if (msg->status() == Message::PROGRESSING) CHPLDEF_TODO();
-  if (msg->status() != Message::PENDING) return true;
+  if (msg->status() != Message::PENDING) return false;
 
-  ctx->message("Handling request '%s' with ID '%s'\n",
+  ctx->message("Handling request '%s' with ID %s\n",
                msg->tagToString(),
                msg->idToString().c_str());
 
-  x = msg->compute(ctx);
+  r = M::compute(ctx, p);
 
-  if (x.isProgressingCallAgain) CHPLDEF_TODO();
+  if (r.isProgressingCallAgain) CHPLDEF_TODO();
 
-  if (x.error != Message::OK) {
-    auto cstr = Message::errorToString(x.error);
+  if (r.error != Message::OK) {
+    auto cstr = Message::errorToString(r.error);
     ctx->message("Request failed with code '%s'\n", cstr);
+    return false;
   }
 
-  if (ctx->logger().level() != Logger::OFF) {
-    auto str = x.result.toString();
-    ctx->message("Request completed with result %s\n", str.c_str());
-  }
+  ctx->message("Request '%s' complete...\n", msg->idToString().c_str());
 
-  return false;
+  return true;
 }
+
+template <typename M, typename P, typename R>
+static bool
+doHandleMessage(Server* ctx, M* msg, const P& p, R& r) {
+  if (msg->isNotification()) return doHandleNotification(ctx, msg, p, r);
+  return doHandleRequest(ctx, msg, p, r);
 }
 
 /** Call the request handler above and then mark completed or failed. */
 #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
   void name__::handle(Server* ctx) { \
-    ComputedResult x; \
-    if (doHandleMessage(ctx, this, x)) { \
-      this->r = std::move(x.result); \
+    ComputedResult r; \
+    if (doHandleMessage(ctx, this, p, r)) { \
+      this->r = std::move(r.result); \
```

Comment:
Slightly confused. Your `doHandleRequest` template, parameterized by `R`, accesses its various fields and assigns to it directly. But here, in this message, you assume that the return value is actually `r.result`. Either there's some nesting confusion about what `R` is (is it a thing _wrapping_ the message response, or is it the whole response?) or your use of `R` is overloaded.

Reply:
`R` is the message result + a few other status fields. Maybe it would help if I renamed the template type to `CR`.

Reply:
It would help